src/core/Object3D.js

```diff
@@ -115,8 +216,95 @@ class Object3D extends EventDispatcher {
 
 		this.userData = {};
 
+		this.updateMatrixSideEffects();
+
+	}
+
+	set matrixAutoUpdate( value ) {
+
+		this.matrixData.matrixAutoUpdate = value;
+
+	}
+
+	get matrixAutoUpdate() {
+
+		return this.matrixData.matrixAutoUpdate;
+
+	}
+
+	set matrixWorldNeedsUpdate( value ) {
+
+		this.matrixData.matrixWorldNeedsUpdate = value;
+
+	}
+
+	get matrixWorldNeedsUpdate() {
+
+		return this.matrixData.matrixWorldNeedsUpdate;
+
+	}
+
+	set matrixWorldAutoUpdate( value ) {
+
+		this.matrixData.matrixWorldAutoUpdate = value;
+
+	}
+
+	get matrixWorldAutoUpdate() {
+
+		return this.matrixData.matrixWorldAutoUpdate;
+
+	}
+
+	updateMatrixSideEffects() {
+
+		if ( this.onBeforeMatrixUpdate !== Object3D.prototype.onBeforeMatrixUpdate ) {
+
+			this.matrixData.updateMatrixWorldBefore = this.onBeforeMatrixUpdate.bind( this );
+
+		}
+
+		if ( this.onAfterMatrixUpdate !== Object3D.prototype.onAfterMatrixUpdate ) {
+
+			this.matrixData.updateMatrixWorldAfter = this.onAfterMatrixUpdate.bind( this );
+
+		}
+
+		if ( this.updateMatrixWorld !== Object3D.prototype.updateMatrixWorld ) {
+
+			console.warn( 'Do not override updateMatrixWorld() in Object3D sub-classes.' );
+			console.warn( 'To extend behaviour on updateMatrixWorld(), use onBeforeMatrixUpdate() and onAfterMatrixUpdate().' );
+
+			// We don't know whether the additional processing should be performed before or after updateMatrixWorld()
+			// so we call it twice.
```

Comment:
I think it could be even thrice sometimes - if `this.updateMatrixWorld()` is called, then it performs calculations in the extended function once when directly called, once when called via `updateMatrixWorldBefore`, and once when called via `updateMatrixWorldAfter`.

Reply:
Agreed, in the case where `updateMatrixWorld()` is called directly on e.g. a SkinnedMesh, this will be invoked 3 times.

This is an extremely rare case, because `updateMatrixWorld()` is usually only invoked on the whole scene, rather than a single object.
https://github.com/mrdoob/three.js/pull/14138#issuecomment-392210879

In any case, will update comment from "twice" to "multile times" so that it is not inaccurate.